sdk/src/beta9/runner/function.py

```diff
@@ -131,87 +125,120 @@ def main(channel: Channel):
         container_id = config.container_id
         container_hostname = config.container_hostname
-        if not task_id:
-            raise RunnerException("Invalid runner environment")
 
         # Start the task
         start_time = time.time()
-        start_task_response: StartTaskResponse = gateway_stub.start_task(
-            StartTaskRequest(task_id=task_id, container_id=container_id)
-        )
-
+        start_task_response = start_task(gateway_stub, task_id, container_id)
         if not start_task_response.ok:
             raise RunnerException("Unable to start task")
 
-        # Kick off monitoring task
         context = FunctionContext.new(config=config, task_id=task_id)
-        monitor_task = thread_pool.submit(
-            _monitor_task,
-            context=context,
-            stub_id=config.stub_id,
-            task_id=task_id,
-            container_id=config.container_id,
-            function_stub=function_stub,
-            gateway_stub=gateway_stub,
-        )
-        futures.thread._threads_queues.clear()
 
-        task_status = TaskStatus.Complete
-        error: Optional[BaseException] = None
-
-        # Invoke function
-        try:
-            get_args_resp: FunctionGetArgsResponse = function_stub.function_get_args(
-                FunctionGetArgsRequest(task_id=task_id)
+        # Start monitoring the task
+        with ThreadPoolExecutor() as thread_pool:
+            monitor_future = thread_pool.submit(
+                _monitor_task,
+                context=context,
+                stub_id=config.stub_id,
+                task_id=task_id,
+                container_id=container_id,
+                function_stub=function_stub,
+                gateway_stub=gateway_stub,
             )
 
-            if not get_args_resp.ok:
-                raise InvalidFunctionArgumentsException
-
-            payload: dict = _load_args(get_args_resp.args)
-            args = payload.get("args") or []
-            kwargs = payload.get("kwargs") or {}
-
-            handler = FunctionHandler()
-            result = handler(context, *args, **kwargs)
-        except BaseException as exc:
-            print(traceback.format_exc())
-            result = error = exc
-            task_status = TaskStatus.Error
-        finally:
-            pickled_result = cloudpickle.dumps(result)
-            set_result_resp: FunctionSetResultResponse = function_stub.function_set_result(
-                FunctionSetResultRequest(task_id=task_id, result=pickled_result)
+            # Invoke the function and handle its result
+            try:
+                result = invoke_function(function_stub, context, task_id)
+            except BaseException as e:
+                handle_task_failure(e, gateway_stub, task_id, container_id, container_hostname)
+                raise
+            finally:
+                monitor_future.cancel()
+
+            # End the task and send callback
+            complete_task(
```

Comment:
Is there no need to handle exceptions on these calls? If we fail to complete the task should we re-try?

Reply:
No, its what I thought too but found out that `end_task_and_send_callback` (which is called by complete_task) has this decorator on it.

```python
@retry_grpc_call(exception_to_check=BaseException, tries=4, delay=5, backoff=2)

```

Reply:
I mean we might still need it, but I don't knwo what we'd do with the error since we can complete the task.

Reply:
Ahh I see